Provide API to run ddl and migration manually (e.g. with offline DB)

rPraml opened the pull request:
This change provides API methods to run DDL and migration manually.
This is useful, if you have a setup where you 
- use TenantMode.DB / SCHEMA  or CATALOG
- expect, that the DB may be not available on startup

**How to use**

If you use a multi-tenancy mode, you will have to iterate over the available tenants and call dbMigration for each tenant. (You'll have to set the tenant in the `currentTenantProvider`, so that the multiTenantDbSupplier will switch to the correct tenant

```java
for (int tenant : tenants) {
    currentTenantProvider.set(tenant)
    db.runMigration();
}
```

Note: Ebean and Ebean-datasource will recover properly, if there are connection problems *after* the first successful connect. But the default configuration does not allow you, to construct a io.ebean.Database instance, when the first connect fails.

If you want to start an ebean server where the datasource could be unreachable, you will have to disable all DB access and checks at startup (datasource.failOnStart=false, skipDataSourceCheck=true, ddl.run=false and migration.run=false)
in this case, ebean will also come up, if the database is not accessible.
This allows you, to show a "Database not available" message when you start your application.

*Sample code*
If you use `io.ebean.datasource` you may register a dataSourceAlert and do the migration when the database comes up the first time. Refer to the "TestServerOffline" or the following code draft, how to use this feature.

```java
class MyDbFactory implements DataSourceAlert {
   boolean initialized = false; 
   public Database create() {
       ...
       dataSourceConfig.setAlert(this);
       ...
       db = DatabaseFactory.create(config)
       if (((ConnectionPool)db.dataSource()).isDataSourceUp()) {
          boot();
       }
       return db;
   }
   void synchronized boot() { // synchronized: dataSourceUp is invoked from different thread
       if (init) return; 
       db.runMigration();
       init = true; // run only once after success
   }
   public void dataSourceUp(final DataSource dataSource) {
       boot(); 
   }
}

rbygrave commented on ebean-core/src/main/java/io/ebeaninternal/api/SpiDdlGenerator.java:
```diff
@@ -8,8 +8,14 @@
+   * <p>
+   * Run based on on property settings for ebean.ddl.run etc.
+   */
+  void runDdl();
```
So both of these methods `runDdl()` `runMigration()` are actually not what I was expecting.  Instead I was expecting `start()` (along the lines of the conversation that took place in the prior issue that was just closed).

There are a number of thoughts which suggest `start()` is still the way to go:
1)  runDdl() should really only be a test concern - we never expect to runDdl in production, only migrations are run in production. 
2) `start` does other things like register with ebean container for clustering support

So at this stage I think this PR isn't what we want. I think we still want `start()` (for programmatic start of the ebean database instance, which could fail due to database/datasource not ready, and could run the migration based on existing config.

I can see that `start()` by itself doesn't support the for loop other tenants, running the migration for each tenant case. Hmmm.

I don't see the change that moves the cluster registration that happens in DefaultContainer to only execute after successful `start()`.  To me we need to sort out `start()` first.

rPraml replied:
Hello @rbygrave I've updated this PR and made the `start()` method part of the API.

Migration/Ddl will run if start() is invoked.

We now need also a `skipStart` in the ServerConfig. I found no other property that may be used here, as `dbOffline` means "really offline" - so no ClusterManager etc will be started.

`start()` is also skipped, if `tenantMode.isDdlEnabled()` is false 

In a multi tenant environment you must call `start()` for each tenant. 

With this PR you can cover the following use cases
- create ebean server, when DB is not yet available, but it is expected to come online later
- do DbMigration (and DdlRun) for each tenant`

I'm not yet sure, if `start()` is the correct term here. I think about
- server.init(); // makes more sense, when iterating over tenants
- config.skipInit
- TenantMode.supportsInit

I would like to give you an outlook on my further ideas: 
1.) DdlGenerator becomes a Plugin: Please take a look at this commit: https://github.com/FOCONIS/ebean/commit/699f7a8449d09c84f73c0a4730a5ba31e0430b1b - this would allow to remove the SpiDdlGenerator completely

2.) DbMigration becomes also a Plugin: Please take a look at these commits in this branch: https://github.com/FOCONIS/ebean/commits/DbMigrationPlugin

Both code-branches are experimental but may affect further decisions. First I want to get this PR discussed & eventually merged

cheers
Roland

rbygrave replied:
I don't think they can become plugins due to ordering ... the ddl ought to run before any/all "plugins" are run etc

rPraml commented on ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:
```diff
@@ -357,13 +357,13 @@ public void initialise() {
       encryptKeyManager.initialise();
     }
     serverCacheManager.enabledRegions(config.getEnabledL2Regions());
+    startQueryPlanCapture();
```
I think we can start the query plan capture here also

rbygrave commented on ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultContainer.java:
```diff
@@ -150,8 +154,6 @@ private void startServer(boolean online, DefaultServer server) {
         clusterManager.registerServer(server);
```
I don't think we should register with clusterManager until the server has started.

rbygrave commented on ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java:
```diff
@@ -246,7 +246,7 @@ private void configureServerPlugins() {
   public void executePlugins(boolean online) {
     if (!config.isDocStoreOnly()) {
-      ddlGenerator.execute(online);
+      ddlGenerator.generateDdl();
```
So DDL used to run before the plugins were executed and now that doesn't happen. I don't think this is correct.

rbygrave commented on ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DdlGenerator.java:
```diff
@@ -67,14 +66,7 @@ public DdlGenerator(SpiEbeanServer server) {
-      this.runDdl = config.isDdlRun();
-      this.ddlAutoCommit = databasePlatform.isDdlAutoCommit();
-    }
+    this.ddlAutoCommit = databasePlatform.isDdlAutoCommit();
```
Yeah, it isn't clear to me what the impact of this change is which makes it hard to merge.  It seems we are doing too many things in this PR when we should instead be doing multiple steps / PR's instead.

rPraml commented:
We use Ebean in multi tenant mode and here DDL/DbMigration has to be completely done by the application, because ebean does neither run DbMigration on start up, nor provide an API to initialize the database for a new tenant.

We have the use case, that tenants are created (and also removed) while the application runs. Our app performs the following steps:
- Admin enters the new DB credentials for the tenant in a web-form, 
- The connection is registered in our MultiTenantDataSourceProvider
- we have a CurrentTenantProvider where we set the new tenant (a ThreadLocal) and run ddl-generation by invoking `server.start()` for this thread.
- it is no option for us to do tenant initializion on ebean startup

> I don't think they can become plugins due to ordering ... the ddl ought to run before any/all "plugins" are run etc

that contradicts itself when a multi-tenant mode is active. So plugins (at least the default ones) have to deal with that situation 

For the "ordering" problem, I've introduced a new lifecycle method `start` in [Plugin](https://github.com/FOCONIS/ebean/blob/ebean-new/ebean-api/src/main/java/io/ebean/plugin/Plugin.java), so that `online` will called once, when the server is created and `start` is called for each tenant. This will work for our use cases.

I also extracted DDL run and DbMigration run in separate plugins in our branch, so I've the following plugins in the classpath

- [DdlPlugin](https://github.com/FOCONIS/ebean/blob/ebean-new/ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DdlPlugin.java) generates in the `execute` method the DDL and applies the DDL in the `start` method (for each tenant)
- [DbMigrationPlugin](https://github.com/FOCONIS/ebean/blob/ebean-new/ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DbMigrationPlugin.java) generates the DbMigration in the `online` method. (for this, I've made DbMigration to be fully configurable from properties. See https://github.com/FOCONIS/ebean/commit/9d1f4194aa2d5a9c248f958bf633115caddc422b)
- [DbRunMigrationPlugin](https://github.com/FOCONIS/ebean/blob/ebean-new/ebean-dbmigration-runner/src/main/java/io/ebeaninternal/dbmigration/run/DbRunMigrationPlugin.java) applies the DDL scripts

This would allow to remove all DdlGenerator and MigrationRun code from DefaultServer
https://github.com/FOCONIS/ebean/commit/699f7a8449d09c84f73c0a4730a5ba31e0430b1b#diff-0804968369579e408eea483a9e582485cd7faf94e5d78dc8f7a918c729962a70
https://github.com/FOCONIS/ebean/commit/208e987ab542f83f377dd2957f0a7895f848f1da#diff-0804968369579e408eea483a9e582485cd7faf94e5d78dc8f7a918c729962a70

So if we introduce more lifecycle methods in the `Plugin` interface we could achieve that we can offload all the DDL stuff from defaultserver.

How would I change the plugin interface:
```java
public interface Plugin {
  // called after Server is constructed. Tasks that can be done here
  // - read configs etc from server
  // - this.server = server
  void configure(SpiServer server);

  // called when the server comes 'online', means
  // - all plugins are configured
  // - you can use all features of the server, like background executor
  // - you can not access the database (at least when multi-tenant mode is used)
  void online(boolean online);
  
  // called when the datasource is available. You may access the database the first time here
  // this is called for each tenant.
  default void start() { };

  // called once, when the server is shut down
  void shutdown();
}
```
Of course, I see the ordering problem. So we can either add an `int order()` method and sort the plugins or add a third lifecycle method e.g. `void initDatabase()`. (or do both). So the initializion workflow could look like this pseudocode
```java
plugins = discoverPluginsViaServiceLocator()
sort(plugins, Plugin::order)

for (plugin : plugins)
   plugin.configure(server)

for (plugin : plugins)
   plugin.online(online)

for (tenant : tenants) // is done by application when multi tenant is active
   server.start()
     // this does effectively
     for (plugin : plugins)
        plugin.initDatabase()  // migration plugins can apply DDL or whatever
     for (plugin : plugins)
        plugin.start()  // can fully access the database

// when server stops
for (plugin : plugins)
   plugin.shutdown()
```
When we add the `default` methods, we would not break the Plugin-API.
We would only introduce the breaking change that the DB is not fully initialized in `online` method (which is also the case when multi-tenant mode is used) - as far as I see the existing plugins will have no problems with that.

@rbygrave I sometimes find it difficult to convey the overall picture and may need some feedback if my text is understandable and what you think about this change. Maybe you can give me some tips, how I should proceed with this PR.

rPraml commented:
Note to myself: This is still on my to-do list ;)

rPraml commented:
This PR is probably obsolete
